### PR TITLE
Feature/valuable integration

### PR DESCRIPTION
**Pull request by xd009642**
Start the initial integration of Valuable into tracing by adding it to the visit trait, implementing Value for Valuable and seeing if it all works. For further motivation see the tracking issue: #1570 

For my own testing, I have a project with the following main.rs, I grabbed the Person and Address types from a valuable example:

```rust
use valuable::Valuable;
use tracing::{info, info_span};

#[derive(Valuable)]
pub struct Person {
    name: String,
    age: u32,
    addresses: Vec<Address>,
}

#[derive(Valuable)]
pub struct Address {
    street: String,
    city: String,
    zip: String,
}

fn blah(value: &(dyn Valuable + 'static)) -> tracing::Span {
    info_span!("span", v=value)
}

fn blahblah(person: &Person) -> tracing::Span {
    info_span!("span", v=person)
}

fn main() {

    let person = Person {
        name: "Daniel".to_string(),
        age: 28,
        addresses: vec![
            Address {
                street: "redacted".to_string(),
                city: "London".to_string(),
                zip: "redacted".to_string()
            }
        ]
    };

    blah(&person);
    blahblah(&person);
}
```

Currently `blah` works, but blahblah has a compilation error which I'm currently trying to resolve but I rarely touch macros in rust so any help would be appreciated:

```
error[E0277]: the trait bound `Person: tracing::Value` is not satisfied
  --> src/main.rs:23:5
   |
23 |     info_span!("span", v=person)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `tracing::Value` is not implemented for `Person`
   |
   = note: required because of the requirements on the impl of `tracing::Value` for `&Person`
   = note: required for the cast to the object type `dyn tracing::Value`
   = note: this error originates in the macro `$crate::valueset` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Person: tracing::Value` is not satisfied
  --> src/main.rs:41:5
   |
41 |     info_span!("Greeting", person=person);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `tracing::Value` is not implemented for `Person`
   |
   = note: required for the cast to the object type `dyn tracing::Value`
   = note: this error originates in the macro `$crate::valueset` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0277`.
error: could not compile `tracing-valuable-test` due to 2 previous errors
```

### CODE REVIEW

**Comment by xd009642** [comment]
Okay after some useful guidance from @hawkw on discord I've added a `ValuableValue` type (candidates for better names more than welcome). And the following minimal example compiles. 

```rust
use valuable::Valuable;
use tracing::{info, info_span};

#[derive(Valuable)]
pub struct Person {
    name: String,
    age: u32,
    addresses: Vec<Address>,
}

#[derive(Valuable)]
pub struct Address {
    street: String,
    city: String,
    zip: String,
}

fn blah(value: &(dyn Valuable + 'static)) -> tracing::Span {
    info_span!("span", v=tracing::field::valuable(value))
}

fn blahblah(person: &Person) -> tracing::Span {
    info_span!("span", v=tracing::field::valuable(person))
}

fn main() {

    let person = Person {
        name: "Daniel".to_string(),
        age: 28,
        addresses: vec![
            Address {
                street: "redacted".to_string(),
                city: "London".to_string(),
                zip: "redacted".to_string()
            }
        ]
    };

    blah(&person);
    info_span!("Greeting", person=tracing::field::valuable(person));
    info!("Hello");
}
```

I haven't yet made any moves to gate it behind a cfg like `RUSTFLAGS="--cfg tracing_unstable"` as suggested in the tracking issue by @davidbarsky. I guess my only question for that would be should it still be an optional feature then or remove the feature entirely in favour of the cfg. Either way that should be a simple change so I figured i'd do that as part of the tidy up once everyone's happy with how it's currently looking and working

**Comment by hawkw** [comment]
>I guess my only question for that would be should it still be an optional feature then or remove the feature entirely in favour of the cfg.

Regardless of whether or not we add an unstable cfg, it should be an optional dependency. The goal is that the `tracing_unstable` cfg would be temporary, and it would eventually be removed. However, once the unstable cfg is removed, I think we would still want users to be able to determine whether or not the `valuable` dependency is enabled. So, if we add the cfg, we should still feature flag as well.

**Comment by hawkw** [comment]
I'm going to try to give this a closer review this afternoon. However, I wanted to start by saying that we are not going to be able to merge this PR until there's a crates.io release of `valuable`. Since we can't publish new `tracing` releases if we have git dependencies, merging this PR would block publishing tracing releases until `valuable` is on crates.io.

Therefore, since this is currently blocked on a published `valuable` release, I'm marking this PR as a draft, since it can't be merged yet for reasons not directly related to the code change here.

cc @carllerche

**Comment by xd009642** [comment]
I've just added the config, missing debug implementation and missing doc comments for this PR. Pending any review comments and a valuable release to crates.io the only thing I can see that would need doing is an update to the CI so the valuable code is also tested. But not sure how thorough the checks should be (i.e. do we still care about MSRV for an unstable feature).

If I have time to do it I'll also work on playing around using this branch on a personal project and if anything of interest comes up or I push anything publicly I'll mention it :smile:

File: tracing-core/src/field.rs
Comment (hawkw) [inline tracing-core/src/field.rs:277]: nit, typos:
```suggestion
/// can be recorded using its `Valuable` implementation.
```

**Review by hawkw — COMMENTED** [review]
Okay, I finally had a chance to give this a review, and overall it looks great. I commented on a few small things that I think we ought to change.

It would definitely also be nice to have some tests for this, and to flesh out the docs and examples somewhat, but that isn't strictly necessary until we can actually merge this PR (which is still blocked on an actual release of `valuable`). So, we can do that later --- for now, I think testing out the branch to make sure it's actually usable is probably the most important next step!

**Comment by xd009642** [comment]
Oh yeah tests would be good, I guess I'll work out something with a custom visitor at some point to make sure the dispatch works correctly - unless there's any smart/more-important things to test with this. When I have some code using this branch I'll apply the unresolved suggested change and see what it's like with and without it

**Comment by hawkw** [comment]
Another good way to test out that this actually works correctly would be writing a few quick examples in the `examples/` dir. That way, we can demonstrate that it works, _and_ when it's ready to merge, we'll already have some nice examples for users to learn how to use it.

**Comment by xd009642** [comment]
Added a very simple example, emphasis on very simple it just derives `Valuable` on a type adds it as a field to a span and then emits an event with the default format subscriber so you can see it printed out. I also cribbed the person/address example from valuable again with a lord of the rings reference because it was the first thing to come to mind that shouldn't covered by copyright wasn't my own name and address :sweat_smile:

Clippy doesn't like the example without a main function though, so maybe attributes on everything is a nicer approach just for CI

**Comment by xd009642** [comment]
@hawkw you might (probably will) be interested in this using the valuable integration branch https://github.com/alsuren/robokite/pull/1

**Comment by hawkw** [comment]
@xd009642 `valuable` 0.1 has [now been released](https://crates.io/crates/valuable) :tada:  so I think we can start to move forwards on getting this merged now!

**Comment by xd009642** [comment]
Awesome I just pushed a quick commit to update to published version before I went offline for the evening, I'll look into it if anything more needs doing later this week :grin:

**Comment by xd009642** [comment]
Moving this out of draft as valuable now released, I've just merged in the v0.1.x branch changes I didn't have. Previously it failed with a security check but not sure what I can do about that? 

Are there any other tasks that need doing for this PR, afaik I've covered all the required feedback. Just the potential solution of the Valuable::as_value implementation of the visitor hasn't been trialled though the comment about it is uncertain if it's the right approach or not. 

Also, not sure if I should change any of the CI workflows to explicitly check the valuable examples?

**Comment by xd009642** [comment]
Oh and a clippy failure which makes no sense to me because the fields must be read when they're printed out!

**Comment by hawkw** [comment]
> Oh and a clippy failure which makes no sense to me because the fields must be read when they're printed out! But I'll figure something later this evening to silence it

I believe the clippy failure is occurring because the fields are only read when compiled with `--cfg tracing_unstable` (in the derived `Valuable` impl). If we compile the example without `tracing_unstable` set, all the fields appear unused.

**Comment by xd009642** [comment]
I put something in much earlier in the PR to avoid that issue though, so I feel this is maybe a macro expansion based issue? Or printing something out doesn't count as reading it :thinking:  

```rust
#[cfg(not(tracing_unstable))]
fn main() {
    tracing_subscriber::fmt()
        .with_max_level(tracing::Level::TRACE)
        .init();

    let user = User {
        name: "Arwen Undomiel".to_string(),
        age: 3000,
        address: Address {
            country: "Middle Earth".to_string(),
            city: "Rivendell".to_string(),
            street: "leafy lane".to_string(),
        },
    };

    let span = info_span!("Processing", user = ?user);
    let _handle = span.enter();
    info!("Nothing to do");
}
```

**Comment by hawkw** [comment]
> I put something in much earlier in the PR to avoid that issue though, so I feel this is maybe a macro expansion based issue? Or printing something out doesn't count as reading it thinking
> 
> ```rust
> #[cfg(not(tracing_unstable))]
> fn main() {
>     tracing_subscriber::fmt()
>         .with_max_level(tracing::Level::TRACE)
>         .init();
> 
>     let user = User {
>         name: "Arwen Undomiel".to_string(),
>         age: 3000,
>         address: Address {
>             country: "Middle Earth".to_string(),
>             city: "Rivendell".to_string(),
>             street: "leafy lane".to_string(),
>         },
>     };
> 
>     let span = info_span!("Processing", user = ?user);
>     let _handle = span.enter();
>     info!("Nothing to do");
> }
> ```

hmm, maybe you're right and the lint somehow misses the derived `Debug` impl? that seems weird.

i think it's fine to just add an `allow` attribute for that warning, for now --- I'd really like to be able to merge this PR!

**Comment by xd009642** [comment]
I'll add an allow dead code then, hopefully CI will pass :pray:

**Comment by xd009642** [comment]
Also I got some extra feedback about alsuren/robokite#1, I could tag you on tokio discord or DM you on discord whichever you prefer :+1:

**Comment by hawkw** [comment]
whatever works!

**Comment by xd009642** [comment]
Okay all CI checks pass :+1:

File: tracing-core/src/field.rs
Comment (hawkw) [inline tracing-core/src/field.rs:286]: might be nice to link to the `Valuable` documentation here?

File: tracing-core/src/field.rs
Comment (hawkw) [inline tracing-core/src/field.rs:261]: might be nice to link to the `Valuable` documentation here?

File: examples/examples/valuable.rs
Comment (hawkw) [inline examples/examples/valuable.rs:66]: would it be possible to unify this so that only the actual use of `valuable` is feature flagged? something like:
```suggestion
fn main() {
    tracing_subscriber::fmt()
        .with_max_level(tracing::Level::TRACE)
        .init();

    let user = User {
        name: "Arwen Undomiel".to_string(),
        age: 3000,
        address: Address {
            country: "Middle Earth".to_string(),
            city: "Rivendell".to_string(),
            street: "leafy lane".to_string(),
        },
    };

    // if the `valuable` feature is enabled, record `user` using its'
    // `valuable::Valuable` implementation:
    #[cfg(tracing_unstable)]
    let span = info_span!("Processing", user = valuable(&user));
    
    // otherwise, record `user` using its `fmt::Debug` implementation:
    #[cfg(not(tracing_unstable))]
    let span = info_span!("Processing", user = ?user);
    
}
```

File: examples/examples/valuable_instrument.rs
Comment (hawkw) [inline examples/examples/valuable_instrument.rs:13]:
```suggestion
    // Current there's no way to automatically apply valuable to a type, so we need to make use of
```

File: examples/examples/valuable.rs
Comment (hawkw) [inline examples/examples/valuable.rs:6]:
```suggestion
//! This example shows how a field value may be recorded using the `valuable` 
//! crate (https://crates.io/crates/valuable). 
//!
//! `valuable` provides a lightweight but flexible way to record structured data, allowing
//! visitors to extract individual fields or elements of structs, maps, arrays, and other 
//! nested structures. 
//!
//! `tracing`'s support for `valuable` is currently feature flagged. Additionally, `valuable`
//! support is considered an *unstable feature*: in order to use `valuable` with `tracing`,
//! the project must be built with `RUSTFLAGS="--cfg tracing_unstable`.
//!
//! Therefore, when `valuable` support is not enabled, this example falls back to using
//! `fmt::Debug` to record fields that implement `valuable::Valuable`.
```

File: examples/examples/valuable.rs
Comment (hawkw) [inline examples/examples/valuable.rs:22]: i think we can have the example always depend on `valuable`, and always derive the trait --- it's specifically *tracing's* `valuable` support that's feature-flagged. that should make this look a little cleaner:
```suggestion
use valuable::Valuable;

#[derive(Valuable, Clone, Debug)]
struct User {
    name: String,
    age: u32,
    address: Address,
}

#[derive(Valuable, Clone, Debug)]
```

**Review by hawkw — APPROVED** [review]
overall, this looks good. I mostly made some documentation and style suggestions.

in follow-up PRs, it would be really nice to:
- add more detailed documentation about how to use `valuable` with tracing in the top-level `tracing_core::field` module docs
- add an example that actually *uses* `valuable` by implementing a visitor or extracting specific values, which we don't currently demonstrate anywhere
- consider adding valuable support to the json formatter for recording nested structs as JSON

File: examples/examples/valuable.rs
Comment (hawkw) [inline examples/examples/valuable.rs:53]: personally, i think it's nice to have comments here explaining the two different lines, like i suggested in my review. not a hard blocker though.

**Review by hawkw — APPROVED** [review]

File: examples/examples/valuable.rs
Comment (xd009642) [reply examples/examples/valuable.rs:53]: Oh sorry fixated on the part about simplifying it to one main function! I'll add a comment now

Comment (xd009642) [reply examples/examples/valuable.rs:57]: Done

File: examples/examples/valuable.rs
Comment (hawkw) [inline examples/examples/valuable.rs:57]:
```suggestion
    // If the `valuable` feature is enabled, record `user` using its'
    // `valuable::Valuable` implementation:
    #[cfg(tracing_unstable)]
    let span = info_span!("Processing", user = valuable(&user));
    
    // Otherwise, record `user` using its `fmt::Debug` implementation:
    #[cfg(not(tracing_unstable))]
    let span = info_span!("Processing", user = ?user);
```

File: tracing-core/Cargo.toml
Comment (bridiver) [inline tracing-core/Cargo.toml:30]: this causes cargo to download the `valuable` crate even though it isn't actually used without `tracing_unstable`

**Pull request merged**
